### DISCUSSION TITLE
Setting up Cloudflare

### DISCUSSION THREAD

**Discussion by poperigby** [troglobit/inadyn#425] [Q&A] [answered]
Hi, I'm trying to setup Inadyn for Cloudflare, but I'm having some trouble. I'm not really sure what the `zone.name` in `username` and `hostname` mean. What should I put there?

**Comment by troglobit** [comment]
I have never used Cloudflare myself, to the best of my knowledge the existing documentation and examples should cover it.  E.g.:

```
# Create a unique custom api token with the following permissions: Zone.Zone - Read, Zone.DNS - Edit.
# ttl is optional, value of 1 is 'automatic'.
# proxied is also optional.
provider cloudflare.com {
    username = zone.name
    password = api_token
    hostname = hostname.zone.name
    ttl      = 1
    proxied  = false
}
```

**Comment by moosefuel** [comment]
> Hi, I'm trying to setup Inadyn for Cloudflare, but I'm having some trouble. I'm not really sure what the `zone.name` in `username` and `hostname` mean. What should I put there?

I also am having trouble getting cloud flare set up. My config might help you. Note that my problem is different: @troglobit, while I can get it to add new records to cloud flare DNS, I cannot get it to update existing ones with the new IP address. 

Here is the config, with just the uncommented stuff (note that I am trying to update several domains in this config, all seem to work if I first delete the A-record with the IP, but if the record exists it doesn't update it):

`# How often the IP is checked. The value denotes seconds
period = 3000

provider cloudflare.com:1 {
    username = moosefuel.media
    password = API_KEY_I_CREATED_ON_CLOUDFLARE_WITH_READ_AND_WRITE
    hostname = moosefuel.media
    ttl = 1 # optional, value of 1 is 'automatic'.
    proxied = false # optional.
}

provider cloudflare.com:2 {
    username = chasingarrows.ca
    password = API_KEY_I_CREATED_ON_CLOUDFLARE_WITH_READ_AND_WRITE
    hostname = chasingarrows.ca
    ttl = 1 # optional, value of 1 is 'automatic'.
    proxied = false # optional.
}

provider cloudflare.com:3 {
    username = aftermath.academy
    password = API_KEY_I_CREATED_ON_CLOUDFLARE_WITH_READ_AND_WRITE
    hostname = aftermath.academy
    ttl = 1 # optional, value of 1 is 'automatic'.
    proxied = false # optional.
}

provider cloudflare.com:4 {
    username = folkl.org
    password = API_KEY_I_CREATED_ON_CLOUDFLARE_WITH_READ_AND_WRITE
    hostname = folkl.org
    ttl = 1 # optional, value of 1 is 'automatic'.
    proxied = false # optional.
}

provider cloudflare.com:5 {
    username = loverabbimovie.com
    password = API_KEY_I_CREATED_ON_CLOUDFLARE_WITH_READ_AND_WRITE
    hostname = loverabbimovie.com
    ttl = 1 # optional, value of 1 is 'automatic'.
    proxied = false # optional.
}

provider cloudflare.com:6 {
    username = shiramovie.com
    password = API_KEY_I_CREATED_ON_CLOUDFLARE_WITH_READ_AND_WRITE
    hostname = shiramovie.com
    ttl = 1 # optional, value of 1 is 'automatic'.
    proxied = false # optional.
}

provider cloudflare.com:7 {
    username = newtfilm.com
    password = API_KEY_I_CREATED_ON_CLOUDFLARE_WITH_READ_AND_WRITE
    hostname = newtfilm.com
    ttl = 1 # optional, value of 1 is 'automatic'.
    proxied = false # optional.
}

`

**Comment by SimonPilkington** [comment]
`zone.name` is the name of the zone as displayed when creating the API token. It will look something like `yourdomain.tld`.

![image](https://github.com/troglobit/inadyn/assets/8479659/5e4ddd53-4c35-4512-ba56-a88a777bf95b)

`hostname` refers to a specific host on your domain that you want to set the address of; this includes the top level, so it can be something like `subdomain.yourdomain.tld` or just `yourdomain.tld`.

**Comment by moosefuel** [comment]
> `hostname` refers to a specific host on your domain that you want to set the address of; this includes the top level, so it can be something like `subdomain.yourdomain.tld` or just `yourdomain.tld`.

Thank you @SimonPilkington, because I need to update multiple domains on the same Cloudflare account, I have mine set to “all zones from a specific user” but I could just as easily use “all zones.” The option to select specific zones is greyed out, maybe I had to set this when I created the token or maybe my account just doesn’t allow it, but I cannot use it for multiple domains anyway.

![IMG_6570](https://github.com/troglobit/inadyn/assets/111374276/42af0668-0669-4ab8-b1eb-e53a0b9a2c0e)

One thing I didn’t realize because of this is that my domains in the config should maybe have the .tld extension, does this apply in the config file for inadyn as well? For example, should mydomain.com be mydomain.com.tdl instead?

**Reply by SimonPilkington** [reply]
>The option to select specific zones is greyed out, maybe I had to set this when I created the token or maybe my account just doesn’t allow it, but I cannot use it for multiple domains anyway.

I couldn't say why that is. I could always select this option on my account.

>One thing I didn’t realize because of this is that my domains in the config should maybe have the .tld extension, does this apply in the config file for inadyn as well? For example, should mydomain.com be mydomain.com.tdl instead?

No, I was using `.tld` as a placeholder, TLD standing for top level domain. In your example that is `.com`, so you would use `mydomain.com`.

**Reply by moosefuel** [reply]
Well it's good to know I was using it right. 

My real problem is that the cloudflare plugin will not update the IP address of any records, but if I create a spurious subdomain, like hi.moosefuel.media (when moosefuel.media is my TLD) it will create it and add the correct IP address, but if I go and change the IP address by one digit, then run inadyn again, it will not update the incorrect IP address to the correct one. Same for TLD. Is there something in my syntax which I am doing wrong?

I just want it to update the IP address of the TLDs I own because when there is a power outage and the battery backup runs out, when it comes back on the ISP assigns a new IP address, and I want it to update automatically, even if I am not at home to fix it.

**Reply by SimonPilkington** [reply]
Your configuration looks correct and everything works as expected for me so I can't say what problem you're having. Try checking debug output as troglobit suggested here: https://github.com/troglobit/inadyn/discussions/400#discussioncomment-6474789

**Reply by moosefuel** [reply]
Will do, thanks.

**Reply by moosefuel** [reply]
So I have had success updating the ip after using this command:

inadyn -1 -l debug --foreground --force

I believe it is because of the --force command. For some reason inadyn doesn’t seem to update it otherwise. I’m still trying to figure out why, looking at the logs.

is there a way to force the update of the ip from the config file? It would be a good interim solution.

**Comment by moosefuel** [comment, marked as answer]
Finally, it appears inadyn was working correctly, it was me that was the problem, and my method of testing.

i assumed that inadyn checks the IP of my TLD against the current IP of the server, but in fact it first checks if the server’s IP has changed since it was last run, and then updates the server.

So in my case to test it I was modifying the IP on Cloudflare and then running inadyn to correct it. Inadyn, however, checked and saw that the server’s IP had not changed, so it never went further unless I applied the --force command. 

To test this I rebooted my modem to get a new IP address, and sure enough, running inadyn corrected it immediately. My problem is solved tentatively for now.

**Reply by troglobit** [reply]
Great to hear you got it working, and thanks for getting back to use with the root cause of your problem.

**Comment by bjackman** [comment]
Not entirely on-topic, but in case anyone is like me and poring over this issue trying to see if they've missed anything: I couldn't get anything to work (it would 403 and say `Zone '<my zone>' not found.`). Eventually I tried just "rolling" the API key in the Cloudflare UI and the new one worked.

I can't see any documentation that confirms this but I suspect the issue was that my original API key contained a hyphen. `man inadyn.conf` has some examples where strings are enclosed in double quotes so maybe that would have fixed it. If this hyphen theory is correct I dunno why I didn't get a parse error.
